src/pybind/mgr/orchestrator/_interface.py

```diff
@@ -748,12 +752,14 @@ def get_feature_set(self):
 
         :returns: Dict of API method names to ``{'available': True or False}``
         """
-        module = self.__class__
-        features = {a: {'available': getattr(Orchestrator, a, None) != getattr(module, a)}
-                    for a in Orchestrator.__dict__
-                    if not a.startswith('_') and not getattr(getattr(Orchestrator, a), '_hide_in_features', False)
-                    }
-        return features
+        with self._lock:
+            if self._features is None:
+                module = self.__class__
+                self._features = {a: {'available': getattr(Orchestrator, a, None) != getattr(module, a)}
+                            for a in Orchestrator.__dict__
+                            if not a.startswith('_') and not getattr(getattr(Orchestrator, a), '_hide_in_features', False)
+                            }
+            return self._features
```

Comment:
Given that this looks like a static getter method (no input arguments and the output seems not to change over time), I was going to suggest using Python [`@functools.cached_property`](https://docs.python.org/3/library/functools.html#functools.cached_property), but that's quite a new addition (not even backported to 3.7) and the only available RPM for CentOS is based on a [different code base](https://github.com/pydanny/cached-property).

Good thing was that it dealt with [`threading.lock` for you](https://github.com/python/cpython/blob/master/Lib/functools.py#L1169) for you.

After a second though I saw that this getter doesn't need to happen at object scope at all. However, it cannot be a 'class constant' either, because it depends on the derived class members (generated metadata) so why not putting this in the `__new__` or `__init__` methods? That way you don't even need to worry about concurrency.

Reply:
Thanks! the `@functools.cached_property` call is new to me and quite handy. But let's stick with what py 3.6 provides since most downstream distros use it...

Doing this in `Orchestrator.__init__` doesn't work (I got a segfault), orchestrator backend treats `Orchestrator` as a mixin class. A backend subclasses `Orchestrator` and `MgrModule`. If I implement `Orchestrator.__init__`, `MgrModule.__init__` is not called according to Python's mro.

`__new__` works, but I need to adjust the inheritance order of some orchestrator backends. Again, because `Orchestrator` is a mixin class, I don't feel right to do this here...

Reply:
> (I got a segfault)

:man_facepalming: Yeah, CPython + Mixins + Reflection sounds like a crazy plan.

>  If I implement Orchestrator.__init__, MgrModule.__init__ is not called according to Python's mro.

Missing `super().__init__()` anywhere?

Nevermind, as @sebastian-philipp if this is not a bottleneck, let's spend no more time on this.

Reply:
Agree. There is no reason to make code more complicated just to optimize this.